### PR TITLE
BUX-380/buxKeysGeneratorContainer

### PR DESCRIPTION
**Pull request by Nazarii-4chain**
Added key generator image

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+		fmt.Println(err)
+	}
+
+	xpubKeyName := "xpub_key.txt"
+	xprvKeyName := "xprv_key.txt"
+
+	xpubKeyFile, err := os.Create(xpubKeyName)
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (dorzepowski) [inline main.go]: It should rather stop the program if there is an error. So for example panic or os.Exit or log.Fatal will be good here.

File: main.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		err := xpubKeyFile.Close()
+		if err != nil {
+			fmt.Println(err)
+		}
+	}(xpubKeyFile)
+
+	_, err = xpubKeyFile.WriteString(keys.XPub().String())
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (dorzepowski) [inline main.go]: same here. it should stop the execution if there is an error

File: main.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+		err := xprvKeyFile.Close()
+		if err != nil {
+			fmt.Println(err)
+		}
+	}(xprvKeyFile)
+
+	_, err = xprvKeyFile.WriteString(keys.XPriv())
+
+	if err != nil {
+		fmt.Println(err)
```
Comment (dorzepowski) [inline main.go]: and here also it should stop the execution

File: main.go [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	fmt.Println("xprv_key.txt created and written successfully.")
+
+	fmt.Println(keys.XPriv())
```
Comment (dorzepowski) [inline main.go]: Don't print that in a console

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+# most recent version of that image when you build your Dockerfile. If
+# reproducability is important, consider using a versioned tag
+# (e.g., alpine:3.17.2) or SHA (e.g., alpine@sha256:c41ab5c992deb4fe7e5da09f67a8804a46bd0592bfdf0b1847dde0e0889d2bff).
+FROM bitnami/kubectl:latest AS final
+
+WORKDIR /src
+
+ENV SECRET_NAME_ENV=bux-keys
+ENV XPUB_KEY_NAME_ENV=admin_xpub
+ENV XPRV_KEY_NAME_ENV=admin_xpriv
```
Comment (dorzepowski) [inline Dockerfile]: Remove the suffix `_ENV` from the variable names

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,63 @@
+# syntax=docker/dockerfile:1
+
+# Comments are provided throughout this file to help you get started.
+# If you need more help, visit the Dockerfile reference guide at
+# https://docs.docker.com/engine/reference/builder/
+
+################################################################################
```
Comment (dorzepowski) [inline Dockerfile]: Remove the comments

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+# Create a new stage for running the application that contains the minimal
+# runtime dependencies for the application. This often uses a different base
+# image from the build stage where the necessary files are copied from the build
+# stage.
+#
+# The example below uses the alpine image as the foundation for running the app.
+# By specifying the "latest" tag, it will also use whatever happens to be the
+# most recent version of that image when you build your Dockerfile. If
+# reproducability is important, consider using a versioned tag
+# (e.g., alpine:3.17.2) or SHA (e.g., alpine@sha256:c41ab5c992deb4fe7e5da09f67a8804a46bd0592bfdf0b1847dde0e0889d2bff).
```
Comment (dorzepowski) [inline Dockerfile]: Remove the comments

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+# Copy the executable from the "build" stage.
+COPY --from=build src/xpub_key.txt .
+COPY --from=build src/xprv_key.txt .
+COPY --from=build src/set_secret.sh .
+
+USER root
+RUN chmod +x /src/set_secret.sh
+USER 1001
+
+# What the container should run when it is started.
```
Comment (dorzepowski) [inline Dockerfile]: Remove the comment

File: Dockerfile [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+# Build the application.
+# Leverage a cache mount to /go/pkg/mod/ to speed up subsequent builds.
+# Leverage a bind mount to the current directory to avoid having to copy the
+# source code into the container.
+RUN --mount=type=cache,target=/go/pkg/mod/ \
+    --mount=type=bind,target=. \
+    CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -v -o /generator
+
+RUN ./generator
```
Comment (dorzepowski) [inline Dockerfile]: This way, you will generate the keys only once during the build, and then everyone who will use that image will have the same values of the keys.

You should rather only build binary at this stage and copy that binary in the next stage and run it as a part of the entrypoint. (so the best would be to have the script that at first will run the binary and then create a secret).

File: compose.yaml
Comment (dorzepowski) [inline compose.yaml:1]: Is this file useful at all? IMHO the kubectl won't work in the docker compose 😕

File: Dockerfile [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+COPY . /src
+
+# Download dependencies as a separate step to take advantage of Docker's caching.
+# Leverage a cache mount to /go/pkg/mod/ to speed up subsequent builds.
+# Leverage bind mounts to go.sum and go.mod to avoid having to copy them into
+# the container.
+RUN --mount=type=cache,target=/go/pkg/mod/ \
+    --mount=type=bind,source=go.sum,target=go.sum \
+    --mount=type=bind,source=go.mod,target=go.mod \
```
Comment (dorzepowski) [inline Dockerfile:11]: hmmm.. I'm not sure about all those mount flags (especially as you're doing copy in a previous step), maybe it's worth to consult them with Marek 😐

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    go mod download -x
+
+# Build the application.
+# Leverage a cache mount to /go/pkg/mod/ to speed up subsequent builds.
+# Leverage a bind mount to the current directory to avoid having to copy the
+# source code into the container.
+RUN --mount=type=cache,target=/go/pkg/mod/ \
+    --mount=type=bind,target=. \
```
Comment (dorzepowski) [inline Dockerfile:15]: same here I'm not sure about those flags

**Review by dorzepowski — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (Nazarii-4chain) [reply main.go]: Done

File: main.go (hunk 2 above)
Comment (Nazarii-4chain) [reply main.go]: Done

File: main.go (hunk 3 above)
Comment (Nazarii-4chain) [reply main.go]: Done

File: main.go (hunk 4 above)
Comment (Nazarii-4chain) [reply main.go]: Removed

File: Dockerfile (hunk 5 above)
Comment (Nazarii-4chain) [reply Dockerfile]: Removed

File: Dockerfile (hunk 6 above)
Comment (Nazarii-4chain) [reply Dockerfile]: Removed

File: Dockerfile (hunk 7 above)
Comment (Nazarii-4chain) [reply Dockerfile]: Removed

File: Dockerfile (hunk 8 above)
Comment (Nazarii-4chain) [reply Dockerfile]: Removed

File: Dockerfile (hunk 9 above)
Comment (Nazarii-4chain) [reply Dockerfile]: Changed to execute when container is running. Thanks

File: compose.yaml
Comment (Nazarii-4chain) [reply compose.yaml:1]: It is for local purposes. It does not exist in image

File: Dockerfile (hunk 10 above)
Comment (Nazarii-4chain) [reply Dockerfile:11]: This helps to speed up builds and installation of packages

**Review by dorzepowski — APPROVED** [review]

**Pull request merged**